### PR TITLE
Output comparison report to a file

### LINKED ISSUE #7: Output to a file
We want to add the possibility of sending the output to a file.
On both formats `MD` and `Json`

### PR DESCRIPTION
**Pull request by jdesiloniz**
Fixes #7 

This PR adds a feature to generate a report file to be saved to disk that contains the results of the benchmarks comparison.

### CODE REVIEW

**Comment by jdesiloniz** [comment]
@fedefernandez @juanpedromoreno sorry for the amount of files, most of them are just copyright dates updates.

File: modules/core/src/main/scala/com/fortysevendeg/hood/utils/FileUtils.scala [hunk 1]
```diff
@@ -47,6 +48,15 @@ object FileUtils {
       Json
     } else Unknown
 
+  def writeFile[F[_]](file: File, contents: String)(
+      implicit S: Sync[F]): F[Either[Throwable, Unit]] = S.delay {
+    Either.catchNonFatal {
+      val writer = new PrintWriter(file)
+      writer.write(contents)
+      writer.close()
```
Comment (BenFradet) [inline modules/core/src/main/scala/com/fortysevendeg/hood/utils/FileUtils.scala]: maybe you could use `S.bracket` to safely release the writer?

File: modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala [hunk 2]
```diff
@@ -199,17 +232,76 @@ object SbtHoodPlugin extends AutoPlugin with SbtHoodDefaultSettings with SbtHood
...
+      currentFile: String,
+      previousBenchmarks: Map[String, Benchmark],
+      currentBenchmarks: Map[String, Benchmark],
+      benchmarkResults: List[BenchmarkComparisonResult]
+  ): List[Benchmark] = {
+    def benchmarkResultMark(benchmarkName: String): String =
+      benchmarkResults
+        .find(_.previous.benchmark.equalsIgnoreCase(benchmarkName))
+        .map(_.icon)
+        .getOrElse("red_circle".emoji.toString())
```
Comment (BenFradet) [inline modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala]: nice, didn't know about this lib :grin:

File: modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala [hunk 3]
```diff
@@ -199,17 +232,76 @@ object SbtHoodPlugin extends AutoPlugin with SbtHoodDefaultSettings with SbtHood
...
+          benchmarksResults)
+
+      val fileContents = outputFileFormat match {
+        case OutputFileFormatJson => collectedBenchmarks.asJson.noSpaces
+        case OutputFileFormatMd   => benchmarkOutput(benchmarksResults, previousFile, currentFile)
+      }
+
+      FileUtils
+        .writeFile(outputPath, fileContents)
+        .map(r => r.leftMap(e => OutputFileError(e.getMessage)))
```
Comment (fedefernandez) [inline modules/plugin/src/main/scala/com/fortysevendeg/hood/plugin/SbtHoodPlugin.scala]: Another option could be to use `EitherT`
```scala
EitherT(FileUtils.writeFile(outputPath, fileContents))
  .leftMap(e => OutputFileError(e.getMessage)
  .value
```

**Comment by jdesiloniz** [comment]
@BenFradet @fedefernandez thanks for your comments! Could you please review again? :D

File: modules/core/src/main/scala/com/fortysevendeg/hood/utils/FileUtils.scala [hunk 4]
```diff
@@ -17,7 +17,8 @@
 package com.fortysevendeg.hood.utils
 
 import cats.effect.{Resource, Sync}
-import java.io.File
+import java.io.{File, PrintWriter}
+import cats.syntax.either._
```
Comment (fedefernandez) [inline modules/core/src/main/scala/com/fortysevendeg/hood/utils/FileUtils.scala]: You could use `cats.implicits._` to have the full syntax

File: modules/core/src/main/scala/com/fortysevendeg/hood/utils/FileUtils.scala [hunk 5]
```diff
@@ -47,6 +48,12 @@ object FileUtils {
       Json
     } else Unknown
 
+  def writeFile[F[_]](file: File, contents: String)(
+      implicit S: Sync[F]): F[Either[Throwable, Unit]] =
+    S.attempt(
+      S.bracket(S.pure(new PrintWriter(file)))(writer => S.pure(writer.write(contents)))(writer =>
```
Comment (fedefernandez) [inline modules/core/src/main/scala/com/fortysevendeg/hood/utils/FileUtils.scala]: I'd use `delay` instead

```scala
S.bracket(S.delay(new java.io.PrintWriter(file)))(w => S.delay(w.write(contents)))(w => S.delay(w.close())).attempt
```

**Comment by jdesiloniz** [comment]
Thanks for the latest suggestion! Merging now :D

**Pull request merged**